LUCENE-8863: enable loading external Kuromoji dictionary

msokolov opened the pull request:
…abling 13-bit ids

I'm sharing this without tests, for comment, while I try to get the tests running. I'm having trouble with

```
    cd ~/workspace/lucene-solr/lucene/analysis/kuromoji; ant test-tools
```

first I had to make this patch, 

```
   <target name="test-tools" depends="compile-tools-tests">
-    <test-macro dataDir="src/tools/test" junit.classpath="tools.test.classpath"/>
+    <test-macro workDir="src/tools/test" junit.classpath="tools.test.classpath"/>
   </target>
```

now I'm getting this error:

```
/home/ANT.AMAZON.COM/sokolovm/workspace/lucene-solr/lucene/common-build.xml:964: Problem: failed to create task or type antlib:com.carrotsearch.junit4:pickseed
Cause: The name is undefined.
```

which all makes me wonder whether test-tools has become obsolete?

rmuir commented on lucene/analysis/kuromoji/src/tools/java/org/apache/lucene/analysis/ja/util/BinaryDictionaryWriter.java:
```diff
@@ -126,7 +132,9 @@ public int put(String[] entry) {
 
     if ((flags & BinaryDictionary.HAS_BASEFORM) != 0) {
-      assert baseForm.length() < 16;
+      if (baseForm.length() >= 16) {
```
this isn't equivalent to the previous assert.

msokolov replied:
Could you elaborate? I do see the exception message is wrong (should say >=), but I think the assertion is correct" `len < 16` iff `!(len >= 16)`

rmuir replied:
I am referring to the exception message. this is one reason why the `assert` works well, such bugs are impossible, just as long as its run with `-ea` which the build file does :)

rmuir commented on lucene/analysis/kuromoji/src/tools/java/org/apache/lucene/analysis/ja/util/BinaryDictionaryWriter.java:
```diff
@@ -35,9 +35,11 @@
   protected final Class<? extends BinaryDictionary> implClazz;
   protected ByteBuffer buffer;
-  private int targetMapEndOffset = 0, lastWordId = -1, lastSourceId = -1;
+  protected int targetMapEndOffset = 0, lastWordId = -1, lastSourceId = -1;
```
what happened here? why is this protected now?

msokolov replied:
my bad - this was left this way from some earlier debugging efforts. I'll undo

msokolov commented:
I posted an updated patch that addresses the comments, and also adds a test that exercises the builder with some test data. In order to test, I needed to add the ability to load an "external" dictionary; this enables loading a system dictionary from an alternate location on the classpath, or from an external file.

rmuir commented:
+1, thank you for taking the time to do these tests. Can we open a followup issue to fix whatever we need to fix so that your tests actually run every time with `ant test`?

I think its an important step so that we can support different dictionary variants and so on, but it doesn't need to block this PR.

rmuir commented on lucene/analysis/kuromoji/build.xml:
```diff
@@ -136,8 +136,8 @@
-  <target name="test-tools" depends="compile-tools-tests">
-    <test-macro dataDir="src/tools/test" junit.classpath="tools.test.classpath"/>
+  <target name="test-tools" depends="install-junit4-taskdef, compile-tools-tests">
+    <test-macro testsDir="${build.dir}/classes/tools-test" workDir="src/tools/test" junit.classpath="tools.test.classpath"/>
```
just an idea for the future: if we are fixing this tooling (e.g. replacing bogus asserts with real checks so it is usable like this PR does), maybe we should simply move all the `tools*` stuff and tests into the regular source tree?

msokolov replied:
yeah, this felt kind of like a dusty corner where things would easily get lost. I'm just learning about the tools and modules concepts in lucene builds. It seems fine to have the builder available in the analysis jar.

mocobeta commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/BinaryDictionary.java:
```diff
@@ -46,13 +50,31 @@
+   * by a filesystem directory and optionally a filename-prefix, or "classpath:", followed by a
+   * classpath location prefix, or just plain "classpath:" to use this class's name as the prefix.
+   */
+  protected BinaryDictionary(String resourcePrefix) throws IOException {
```
If the resourcePrefix parameter takes only "classpath:" or "file:", would it be possible to use enum here instead of arbitrary String (for safety)?.

mocobeta replied:
And just wondering, why do we need "file:" prefix? I feel like that loading the dictionary data from arbitrary location could be problematic... (Am I missed anything?)

msokolov replied:
enum is a good idea; I'll update the patch. I added "file:" to enable unit testing. We can't load a test dictionary from the classpath, I think. Another possibility would be to pass in dictionaries directly somehow. EG we could maybe have {{...DictionaryBuilder}} {{FST.save()}} to an in-memory buffer, but this seemed like a more intrusive change since we can already save to a set of files.

mocobeta replied:
I missed the tests. OK, thanks!

msokolov commented:
Sorry for all the close/open/close PR noise. I was fiddling around because I don't seem to have permission to push from github, although I have "apache member" on my profile, and I can push from command line (with login prompt)
